idea/idea-fir/src/org/jetbrains/kotlin/idea/fir/intentions/HLConvertToBlockBodyIntention.kt

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.fir.intentions
+
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.idea.KotlinBundle
+import org.jetbrains.kotlin.idea.api.applicator.HLApplicatorInput
+import org.jetbrains.kotlin.idea.api.applicator.applicator
+import org.jetbrains.kotlin.idea.fir.api.AbstractHLIntention
+import org.jetbrains.kotlin.idea.fir.api.applicator.HLApplicabilityRange
+import org.jetbrains.kotlin.idea.fir.api.applicator.HLApplicatorInputProvider
+import org.jetbrains.kotlin.idea.fir.api.applicator.inputProvider
+import org.jetbrains.kotlin.idea.fir.applicators.ApplicabilityRanges
+import org.jetbrains.kotlin.idea.formatter.adjustLineIndent
+import org.jetbrains.kotlin.idea.frontend.api.KtAnalysisSession
+import org.jetbrains.kotlin.idea.frontend.api.types.KtClassErrorType
+import org.jetbrains.kotlin.idea.frontend.api.types.KtType
+import org.jetbrains.kotlin.idea.util.resultingWhens
+import org.jetbrains.kotlin.idea.util.setType
+import org.jetbrains.kotlin.name.ClassId
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.endOffset
+import org.jetbrains.kotlin.psi.psiUtil.startOffset
+
+class HLConvertToBlockBodyIntention :
+    AbstractHLIntention<KtDeclarationWithBody, HLConvertToBlockBodyIntention.Input>(KtDeclarationWithBody::class, applicator) {
+
+    class Input(
+        val returnType: KtType,
+        val returnTypeIsUnit: Boolean,
+        val returnTypeIsNothing: Boolean,
+        val returnTypeString: String,
+        val returnTypeClassId: ClassId?,
+        val bodyTypeIsUnit: Boolean,
+        val bodyTypeIsNothing: Boolean,
+        val reformat: Boolean,
+    ) : HLApplicatorInput
+
+    override fun allowCaretInsideElement(element: PsiElement) = element !is KtDeclaration && super.allowCaretInsideElement(element)
```

Comment:
OK, let me add a doc string for `org.jetbrains.kotlin.idea.intentions.SelfTargetingIntention#allowCaretInsideElement`. Ideally we should probably rename the method, but I am not sure if it's gonna break plugins. @darthorimar can we rename this method?

Reply:
Having read the doc string and looked at the code more closely, should this be just `element !is KtDeclarationWithBody` without the `super` call? Say you have a function like this:
```kt
fun returnsLambda() = { 1 }
```
If you had the cursor inside the `{ 1 }` (a `KtBlockExpression`), then `super.allowCaretInsideElement(element)` would return `false` and you would not get to `returnsLambda`.

I think it should be `KtDeclarationWithBody` (not just `KtDeclaration`), if you have a case like this:
```kt
fun returnsLambda2() = {
    fun foo() = 1
    val x = foo()
    x
}
```
...and you have the cursor inside `val x = foo()` (a `KtProperty` which is a `KtDeclaration`) then you would stop there instead of getting to `returnsLambda2`.

Looks like this was copied from `ConvertToBlockBodyIntention` but I think we can improve on that.

Reply:
I actually think the current behavior is better. Doing `element !is KtDeclarationWithBody` would create this intention in too many places, which can be confusing for user: Why is this convert to block intention offered for this local property?
The same could apply to `fun returnsLambda = { 1 }`. User might think: `{ 1 }` is already a block, why is the IDE offer me to convert it to a block again?

Reply:
Ok fair enough! The user can still put the cursor just before the left-brace or after the right-brace anyway.

Reply:
Renaming `SelfTargetingIntention#allowCaretInsideElement` may break binary compatibility of 3rd party plugins. But you may introduce a new method with the name you like, delegate it to `allowCaretInsideElement` and mark `allowCaretInsideElement` as deprecated